src/wast-parser.cc

```diff
@@ -0,0 +1,2013 @@
+/*
+ * Copyright 2017 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "wast-parser.h"
+
+#include "binary-reader.h"
+#include "binary-reader-ir.h"
+#include "cast.h"
+#include "error-handler.h"
+#include "wast-parser-lexer-shared.h"
+
+#define WABT_TRACING 0
+#include "tracing.h"
+
+#define CHECK_RESULT(expr)  \
+  do {                      \
+    if (Failed(expr))       \
+      return Result::Error; \
+  } while (0)
+
+#define EXPECT(token_type) CHECK_RESULT(Expect(TokenType::token_type))
```

Comment:
I made this a macro mostly because of the magic automatic control flow (since `CHECK_RESULT` will return on failure)